core/server/api_container/server/service_network/default_service_network.go

```diff
@@ -679,43 +683,158 @@ func (network *DefaultServiceNetwork) RemoveService(
-		return stacktrace.Propagate(err, "Failed to pause service '%v'", serviceIdentifier)
+	for erroredUuid, err := range failedServices {
+		erroredUuids[erroredUuid] = err
+	}
+
+	return successfulUuids, erroredUuids, nil
+}
+
+func (network *DefaultServiceNetwork) StopService(
+	ctx context.Context,
+	serviceIdentifier string,
+) error {
+	serviceIdentifiers := []string{serviceIdentifier}
+	_, erroredUuids, err := network.StopServices(ctx, serviceIdentifiers)
+	if err != nil {
+		return stacktrace.Propagate(err, "An error occurred while stopping services")
 	}
+
+	for serviceUuid, erroredUuid := range erroredUuids {
+		return stacktrace.Propagate(erroredUuid, "An error occurred while stopping service '%v'", serviceUuid)
+	}
+
 	return nil
 }
 
+func (network *DefaultServiceNetwork) StopServices(
+	ctx context.Context,
+	serviceIdentifiers []string,
+) (
+	map[service.ServiceUUID]bool,
+	map[service.ServiceUUID]error,
+	error,
+) {
+	network.mutex.Lock()
+	defer network.mutex.Unlock()
+
+	serviceUuids := map[service.ServiceUUID]bool{}
+	serviceRegistrations := map[service.ServiceUUID]*service.ServiceRegistration{}
+
+	for _, serviceIdentifier := range serviceIdentifiers {
+		serviceRegistration, err := network.getServiceRegistrationForIdentifierUnlocked(serviceIdentifier)
+		if err != nil {
+			return nil, nil, stacktrace.Propagate(err, "An error occurred while getting service registration for identifier '%v'", serviceIdentifier)
+		}
+		if serviceRegistration.GetStatus() == service.ServiceStatus_Stopped {
+			return nil, nil, stacktrace.NewError("Service '%v' is already stopped", serviceRegistration.GetName())
+		}
+		serviceUuids[serviceRegistration.GetUUID()] = true
+		serviceRegistrations[serviceRegistration.GetUUID()] = serviceRegistration
+	}
+
+	stopServiceFilters := &service.ServiceFilters{
+		Names:    nil,
+		UUIDs:    serviceUuids,
+		Statuses: nil,
+	}
+	successfulUuids, erroredUuids, err := network.kurtosisBackend.StopUserServices(ctx, network.enclaveUuid, stopServiceFilters)
+	if err != nil {
+		return successfulUuids, erroredUuids, stacktrace.Propagate(err, "An error occurred during the call to stop services")
+	}
+
+	for successfulUuid := range successfulUuids {
+		serviceRegistrations[successfulUuid].SetStatus(service.ServiceStatus_Stopped)
+	}
+
+	return successfulUuids, erroredUuids, nil
+}
+
 // TODO we could switch this to be a bulk command; the backend would support it
-func (network *DefaultServiceNetwork) UnpauseService(ctx context.Context, serviceIdentifier string) error {
+func (network *DefaultServiceNetwork) PauseService(ctx context.Context, serviceIdentifier string) error {
 	network.mutex.Lock()
 	defer network.mutex.Unlock()
 
-	serviceName, err := network.getServiceNameForIdentifierUnlocked(serviceIdentifier)
+	serviceRegistration, err := network.getServiceRegistrationForIdentifierUnlocked(serviceIdentifier)
```

Comment:
Refactor this common operation where we get the service registration for a service identifier.